### PR TITLE
Add UDP echo server example

### PR DESCRIPTION
**Pull request by Licenser**
This adds a simple UDP echo server. It doesn't cover the `edge` vs `level` case since that seems to be gone in current master.

### CODE REVIEW

File: examples/udp_server.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+use mio::{Events, Interests, Poll, Token};
+use std::io;
+
+const IN: Token = Token(0);
+
+fn main() {
+    // Set up a new socket on port 9000 to listen on.
+    let socket = UdpSocket::bind("0.0.0.0:9000".parse().unwrap()).unwrap();
+    // Initialize poller.
+    let mut poll = Poll::new().unwrap();
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Same as above: `unwrap` -> `?`.

File: examples/udp_server.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Prepare a buffer for the number of events we can handle at a time.
+    // Someone might wat to echo really fast so lets give it some size.
+    let mut events = Events::with_capacity(1024);
+    // Initialize a buffer for the UDP datagram
+    let mut buf = [0; 65535];
+    // Main loop
+    loop {
+        // Poll if we have events waiting for us on the socket.
+        poll.poll(&mut events, None).unwrap();
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Same as above: `unwrap` -> `?`.

File: examples/udp_server.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+fn main() {
+    // Set up a new socket on port 9000 to listen on.
+    let socket = UdpSocket::bind("0.0.0.0:9000".parse().unwrap()).unwrap();
+    // Initialize poller.
+    let mut poll = Poll::new().unwrap();
+    // Register our socket with the token IN (defined above) and an interest
+    // in being `READABLE`.
+    poll.registry()
+        .register(&socket, IN, Interests::READABLE)
+        .unwrap();
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Same as above: `unwrap` -> `?`.

File: examples/udp_server.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
+use mio::net::UdpSocket;
+use mio::{Events, Interests, Poll, Token};
+use std::io;
+
+const IN: Token = Token(0);
+
+fn main() {
+    // Set up a new socket on port 9000 to listen on.
+    let socket = UdpSocket::bind("0.0.0.0:9000".parse().unwrap()).unwrap();
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Can you replace the `unwrap` with `?`, returning an `io::Result` from `main`?

File: examples/udp_server.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            // in this example it will only ever be one but we
+            // make sure it's valid non the less.
+            match event.token() {
+                IN => loop {
+                    // In this loop we receive from the socket as long as we
+                    // can read data
+                    match socket.recv_from(&mut buf) {
+                        Ok((n, from_addr)) => {
+                            // Send the data right back from where it came from.
+                            let _r = socket.send_to(&buf[..n], from_addr);
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Can you handle the error here (by using `?`).

File: examples/udp_server.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                            // If we failed to receive data we have two cases
+                            if e.kind() == io::ErrorKind::WouldBlock {
+                                // If the reason was `WouldBlock` we know
+                                // our socket has no more data to give so
+                                // we can return to the poll to wait politely.
+                                break;
+                            } else {
+                                // If it was any other kind of error, something
+                                // went wrong and we panic.
+                                panic!("error: {:?}", e)
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: If `main` is changed to return an `io::Error` it can be returned here.

**Comment by Licenser** [comment]
👍 will update in a bit

File: examples/udp_server.rs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                            }
+                        }
+                    }
+                },
+                // We only have IN as a token, so this should never ever be hit
+                _ => unreachable!(),
+            }
+        }
+    }
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: Add some spaces here.

File: examples/udp_server.rs (hunk 1 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

File: examples/udp_server.rs (hunk 3 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

File: examples/udp_server.rs (hunk 5 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

File: examples/udp_server.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    let mut buf = [0; 65535];
+    // Main loop
+    loop {
+        // Poll if we have events waiting for us on the socket.
+        poll.poll(&mut events, None)?;
+        // If we do iterate throuigh them
+        for event in events.iter() {
+            // Validate the token we registered our socket with,
+            // in this example it will only ever be one but we
+            // make sure it's valid non the less.
```
Comment (Thomasdezeeuw) [inline examples/udp_server.rs]: type-o: non*e* the less.

File: examples/udp_server.rs (hunk 6 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

File: examples/udp_server.rs (hunk 7 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

File: examples/udp_server.rs (hunk 8 above)
Comment (Licenser) [reply examples/udp_server.rs]: 👍

**Comment by Thomasdezeeuw** [comment]
Thanks.

**Comment by Licenser** [comment]
thanks for helping me getting it sorted out in the first place :)

**Pull request merged**